src/app/ipfs-cache.ts

```diff
@@ -27,38 +27,103 @@ let lastCache: IPFSCache;
+   *   => http://crocodillon.com/blog/always-catch-localstorage-security-and-quota-exceeded-errors
+   *
+   * @param      {Exception}   ex      error object
+   * @return     {boolean}  True if quota exceeded, False otherwise
    */
-  constructor(dbName = 'ipfs-cache', readonly storeName = 'ipfs-hashes') {
-    this._dbp = new Promise((resolve, reject) => {
-      const openreq = indexedDB.open(dbName, 1);
-      openreq.onerror = () => {
-        // reject(openreq.error);
-        resolve();
-      };
-      openreq.onsuccess = () => {
-        try {
-          openreq.result.createObjectStore(storeName);
-        } catch (ex) { }
+  static isQuotaExceeded(ex: any): boolean {
+    let quotaExceeded = false;
 
-        resolve(openreq.result);
-      };
+    if (ex) {
+      if (ex.code) {
+        switch (ex.code) {
+          case 22: {
+            quotaExceeded = true;
+            break;
+          }
+          case 1014: {
+            // Firefox
+            if (ex.name === 'NS_ERROR_DOM_QUOTA_REACHED') {
+              quotaExceeded = true;
+            }
 
-      openreq.onblocked = function (event) {
-        log('IndexDB blocked!');
+            break;
+          }
+          default: {
+            // do nothing
+          }
+        }
+      } else if (ex.number === -2147024882) {
+        // Internet Explorer 8
+        quotaExceeded = true;
+      }
+    }
 
-        resolve();
-      };
+    return quotaExceeded;
+  }
 
-      // First time setup: create an empty object store
-      openreq.onupgradeneeded = () => {
-        openreq.result.createObjectStore(storeName);
-      };
-    });
+  /**
+   * Return the currently opened db or open a new one. If no db was opened, before, open it.
+   *
+   * @param      {boolean}  forceNew  open a new one
+   */
+  private async getOpenedDb(forceNew = false): Promise<any> {
+    if (!this.db || forceNew) {
+      this.db = new Promise((resolve, reject) => {
+        log(`[ipfs-cache] open cache db - ${this.dbName}`);
+        const openreq = indexedDB.open(this.dbName, 1);
+        openreq.onsuccess = (): void => {
+          try {
+            openreq.result.createObjectStore(this.storeName);
+          } catch (ex) {
+            // already exists
+          }
+
+          resolve(openreq.result);
+        };
+        openreq.onblocked = (): void => reject();
+        openreq.onerror = (): void => reject();
+
+        // First time setup: create an empty object store
+        openreq.onupgradeneeded = () => {
+          log(`[ipfs-cache] create cache db - ${this.dbName}`);
+          openreq.result.createObjectStore(this.storeName);
+        };
+      });
+    }
+
+    return this.db;
+  }
+
+  /**
+   * Close the current db, delete the previous one and create a new one.
+   */
+  private async recreateDb(): Promise<void> {
+    log('[ipfs-cache] recreate ipfs db - maybe aborted?');
+    // close previous db
+    await (await this.getOpenedDb()).close();
```

Comment:
shall we check beforehand if `this.db` exists, that we don't open a DB just to close it?